fedimint-core/src/db/mod.rs

```diff
@@ -1841,28 +1877,117 @@ pub async fn apply_migrations(
-                migration(&mut dbtx.to_ref_nc()).await?;
+                if let Some(module_instance_id) = module_instance_id {
+                    migration(
+                        &mut global_dbtx
+                            .to_ref_with_prefix_module_id(module_instance_id)
+                            .into_nc(),
+                    )
+                    .await?;
+                } else {
+                    migration(&mut global_dbtx.to_ref_nc()).await?;
+                }
             } else {
                 panic!("Missing migration for version {current_db_version}");
             }
 
             current_db_version.increment();
-            dbtx.insert_entry(&DatabaseVersionKey, &current_db_version)
+            global_dbtx
+                .insert_entry(
+                    &DatabaseVersionKey(module_instance_id_key),
+                    &current_db_version,
+                )
                 .await;
         }
 
         current_db_version
     } else {
-        dbtx.insert_entry(&DatabaseVersionKey, &target_db_version)
-            .await;
         target_db_version
     };
 
-    dbtx.commit_tx_result().await?;
+    global_dbtx.commit_tx_result().await?;
     info!(target: LOG_DB, "{} module db version: {}", kind, db_version);
     Ok(())
 }
 
+/// Migrates the `DatabaseVersion` from inside the module's isolated database
+/// to the global namespace. Database migrations are driven from outside of the
+/// modules so the `DatabaseVersion` should also exist outside the modules.
+/// This also fixes the key prefix conflicts with 0x50.
+pub async fn migrate_database_version(
+    global_dbtx: &mut DatabaseTransaction<'_>,
+    target_db_version: DatabaseVersion,
+    module_instance_id: Option<ModuleInstanceId>,
+    kind: String,
+) -> Result<(), anyhow::Error> {
+    let key_module_instance_id = module_instance_id_or_global(module_instance_id);
+
+    // First check if the module has a `DatabaseVersion` written to
+    // `DatabaseVersionKey` If `DatabaseVersion` already exists, there is
+    // nothing to do.
+    if global_dbtx
+        .get_value(&DatabaseVersionKey(key_module_instance_id))
+        .await
+        .is_none()
+    {
+        // Read and remove the previous `DatabaseVersion`, which is in the module's
+        // isolated namespace (but not for fedimint-server)
+        let current_version_in_module = if let Some(module_instance_id) = module_instance_id {
+            remove_current_db_version_if_exists(
+                &mut global_dbtx
+                    .to_ref_with_prefix_module_id(module_instance_id)
+                    .into_nc(),
+                target_db_version,
+            )
+            .await
+        } else {
+            remove_current_db_version_if_exists(
+                &mut global_dbtx.to_ref().into_nc(),
+                target_db_version,
+            )
+            .await
+        };
```

Comment:
nit: The actual fn call could be factored out, only the DB argument needs to change depending on the instance id being set or not.